### PR TITLE
ILIAS 8: Migration step for `InnoDB`

### PR DESCRIPTION
**Pull request by gvollbach**
Replacement for https://github.com/ILIAS-eLearning/ILIAS/pull/3704
@chfsx Could you please test this also? Thx!

### CODE REVIEW

File: Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                WHERE ENGINE != %s AND table_schema = %s;", ['text', 'text'], [
+                ilDBConstants::MYSQL_ENGINE_INNODB,
+                $this->db_name
+            ]);
+            $row = $this->database->fetchAssoc($set);
+            return (int) $row['tables'];
+        }
+        return 0;
+    }
+}
```
Comment (mjansenDatabay) [inline Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php]: @gvollbach Please use the `PHP-CS-FIXER`.

Comment (gvollbach) [reply Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php]: Done.

File: Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        if ($this->db_name !== null) {
+            $set = $this->database->queryF("SELECT count(*) as tables
+                FROM INFORMATION_SCHEMA.TABLES
+                WHERE ENGINE != %s AND table_schema = %s;", ['text', 'text'], [
+                ilDBConstants::MYSQL_ENGINE_INNODB,
+                $this->db_name
+            ]);
+            $row = $this->database->fetchAssoc($set);
+            return (int) $row['tables'];
```
Comment (chfsx) [inline Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php]: We have a 'int' amount of tables here which we want to migrate, right? Then I think we should consider that in `step()` as well. as far as I see all tables get migrated in one step in `step()` currently?

File: Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        $this->db_name = $client_ini->readVariable('db', 'name');
+    }
+
+    /**
+     * @inheritDoc
+     * @throws ilException
+     */
+    public function step(Environment $environment) : void
+    {
+        $errors = $this->database->migrateAllTablesToEngine();
```
Comment (chfsx) [inline Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php]: we should migrate one by one here, maybe we introduce another method for that in `ilDBInterface` for this

File: Services/Database/classes/class.ilDBConstants.php [hunk 4]
```diff
@@ -85,7 +83,6 @@ class ilDBConstants
         ilDBConstants::TYPE_POSTGRES => "Postgres (experimental)",
         ilDBConstants::TYPE_GALERA => "Galera-Cluster (experimental)",
         // Development identifiers (will be removed in 5.3)
-        ilDBConstants::TYPE_PDO_MYSQL_MYISAM => "MySQL 5.5.x or higher (MyISAM engine) [developers-identifier]",
         ilDBConstants::TYPE_PDO_MYSQL_INNODB => "MySQL 5.5.x or higher (InnoDB engine) [developers-identifier]",
```
Comment (chfsx) [inline Services/Database/classes/class.ilDBConstants.php]: we can change this like this:
```suggestion
        ilDBConstants::TYPE_PDO_MYSQL_INNODB => "MySQL 5.7.x or higher (InnoDB engine)",
```

**Review by chfsx — CHANGES_REQUESTED** [review]
Hi @gvollbach 
thank you for your adjustments! could you add the points in the comments? if i can be of any help, contact me anytime

File: Services/Database/classes/class.ilDBConstants.php (hunk 4 above)
Comment (gvollbach) [reply Services/Database/classes/class.ilDBConstants.php]: Done.

File: Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php (hunk 2 above)
Comment (gvollbach) [reply Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php]: @chfsx 
Yes, that ist the int amount of tables that have not the correct engine, which is needed for the setup, so it knows that there are migration steps todo or left.

File: Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php (hunk 3 above)
Comment (gvollbach) [reply Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php]: @chfsx 
Then we would need a new function in the DBInterface as well as in the other classes, should I implement this, or will you implement this functions?

Comment (chfsx) [reply Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php]: I implemented this but could not make a PR to DatabayAG repo: https://github.com/srsolutionsag/ILIAS/tree/feature/8/migrate-single-table
Can you merge this locally?

Comment (gvollbach) [reply Services/Database/classes/Setup/class.ilMysqlMyIsamToInnoDbMigration.php]: @chfsx 
Thx! And done.

**Pull request merged**